ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/modules/ble_advertising.py

```diff
@@ -0,0 +1,93 @@
+# Helpers for generating BLE advertising payloads.
```

Comment:
This file looks like a copy of that in `examples/bluetooth`.  It would be better to include it in the manifest by referring to that file there, instead of copying it.

But note that no other BLE-enabled board includes this file.  What's the reason to include it here?  Related to that, did you see aioble https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble?  That's a very nice, user friendly asyncio-based wrapper around the `bluetooth` module, and maybe it's worth including that instead.

Reply:
Yes it's an exact copy, and it's frozen so that examples like `examples/bluetooth/ble_temperature.py` work out of the box (without having to somehow copy `ble_advertising.py` to the filesystem). I'll try to add it to the manifest.

Reply:
If you have the flash space I would definitely prefer to include `aioble` instead. For a user of this board, it's a far more useful library, and also specifically for advertising it includes more payload fields.

You can add it to the manifest with

```py
include("$(MPY_LIB_DIR)/micropython/bluetooth/aioble/manifest.py", client=True, central=True, l2cap=True, security=True)
```

Then a user can directly run https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/examples/temp_sensor.py as an alternative to `examples/bluetooth/ble_temperature.py`.

Reply:
@jimmo Yes it has 16MBytes of flash, I just have minor concerns about porting another blinky example based on the temp example, but anyway that path doesn't exist ?

Reply:
I just realized it's in another repo, but how can I link it in the board manifest  ?

Reply:
> I just realized it's in another repo, but how can I link it in the board manifest ?

Exactly as above (I copied it from a board definition I use for testing). The expectation is that the micropython-lib repo is cloned as a sibling directory of the micropython repo (or MPY_LIB_DIR is set otherwise).

The build server is set up like this and e.g. see ports/esp32/boards/manifest_release.py which does something similar for urequests.

Reply:
Thanks, I managed to test it but it's definitely not as stable as the non-aio one, I get disconnected randomly after a while and can't reconnect. Is this on purpose ? The example seems to be running in a while True.

![image](https://user-images.githubusercontent.com/4498013/141647420-2413aeee-b1ce-410b-8c2d-841047dbba2e.png)

Side note, not important but I noticed it's missing something the indicate property ? which I think allows me to auto-pull the value like with the other ble library/example (using nrf connect for testing, see the screenshot, this is the missing option).

![image](https://user-images.githubusercontent.com/4498013/141647106-ca0b3b32-97b6-4a7c-9f48-1c106b7d0f33.png)

Reply:
Interesting... we've been using `aioble` elsewhere (on pybd and stm32-with-external-hci) without issues. I have ordered a Nano 2040 and will investigate when it arrives this week. Does it stop advertising entirely, or just no longer allow connection?

> Side note, not important but I noticed it's missing something the indicate property ?

Yeah, the modbluetooth example sent notifications every 10 seconds. (It has the code to send indications but it's not enabled).

In the `aioble` example, add `True` as a second argument to characteristic.write (i.e. `temp_characteristic.write(_encode_temperature(t), True)`) to automatically notify subscribed clients on write. That should probably be the default in that example, I will send a PR tomorrow.

Reply:
> Does it stop advertising entirely, or just no longer allow connection?

It just no longer allows connections until I hit ctrl+c and run the script again (paste, ctrl-d etc..)

Reply:
We'll definitely fix any issues `aioble` has.  And the goal would be to promote `aioble` and its examples as the recommended way to use bluetooth.  So I think it makes sense to freeze in `aioble` now instead of `ble_advertising.py`.

Reply:
Okay, switched to aioble.